src/shared/firewall-util-nft.c

```diff
@@ -0,0 +1,853 @@
+        int r;
+
+        r = nft_new_set(ret, set_name, flags | NFT_SET_MAP, type, klen);
+        if (r < 0)
+                return r;
+
+        m = *ret;
+        r = sd_netlink_message_append_u32(m, NFTA_SET_DATA_TYPE, htobe32(dtype));
+        if (r < 0)
+               return r;
+
+        return sd_netlink_message_append_u32(m, NFTA_SET_DATA_LEN, htobe32(dlen));
+}
+
+static int nft_add_element(sd_netlink_message **m, const char *set_name,
+                           const void *key, uint32_t klen,
+                           const void *data, uint32_t dlen)
+{
+        int r;
+
+        /*
+         * Ideally there would be an API that provides:
+         *
+         * 1) a init function to add the main ruleset skeleton
+         * 2) a function that populates the sets with all known address/port pairs to s/dnat for
+         * 3) a function that can remove address/port pairs again.
+         *
+         * At this time, the existing API is used which is built on a
+         * 'add/delete a rule' paradigm.
+         *
+         * This replicated here and each element gets added to the set
+         * one-by-one.
+         */
+        r = sd_nfnl_nft_message_new_setelems_begin(nfnl, m, nft_systemd_table_name, set_name);
+        if (r < 0)
+               return r;
+
+        r = sd_nfnl_nft_message_add_setelem(*m, 0, key, klen, data, dlen);
+        if (r < 0)
+               return r;
+
+        /* could theoretically append more set elements to add here */
+        return sd_nfnl_nft_message_add_setelem_end(*m);
+}
+
+static int nft_del_element(sd_netlink_message **m, const char *set_name,
+                           const void *key, uint32_t klen,
+                           const void *data, uint32_t dlen)
+{
+        int r;
+
+        r = sd_nfnl_nft_message_del_setelems_begin(nfnl, m, nft_systemd_table_name, set_name);
+        if (r < 0)
+               return r;
+
+        r = sd_nfnl_nft_message_add_setelem(*m, 0, key, klen, data, dlen);
+        if (r < 0)
+               return r;
+
+        return sd_nfnl_nft_message_add_setelem_end(*m);
+}
+
+/* This is needed so 'nft' userspace tool can properly format the contents
+ * of the set/map when someone lists their contents.
+ */
+enum nft_key_types {
+        TYPE_IPADDR = 7,
+        TYPE_IP6ADDR = 8,
+        TYPE_INET_PROTOCOL = 12,
+        TYPE_INET_SERVICE = 13,
+};
+
+/* this allows nft to format the map_port_ipport concatenated set */
+static uint32_t concat_types2(enum nft_key_types a, enum nft_key_types b)
+{
+        return a << 6 | b;
+}
+
+#define NFT_INIT_MSGS	16
+static int __fw_nftables_init(sd_netlink_message *ruleset[], unsigned int msgcnt)
+{
+        int r;
+
+        /* Set F_EXCL so table add fails if the table already exists. */
+        r = sd_nfnl_nft_message_new_table(nfnl, &ruleset[++msgcnt], nft_systemd_table_name, NLM_F_EXCL | NLM_F_ACK);
+        if (r < 0)
+                return r;
+
+        assert_return(msgcnt < NFT_INIT_MSGS, -EINVAL);
```

Comment:
hmm, assert_return() we use for APIs that are user facing, i.e. that either already are export to be linked to by other programs that use libsystemd, or are written so that they eventually are. sd-netlink is one of those libs (which is supposed to be exposed to clients eventually, but currently is still private). For stuff that is not user-facing, but facing only to our own code we just use plain assert(). Both shall be used only for programming errors (and not runtime errors), the only difference is that we are a bit stricter for ourselves then for users, hence if we encounter a programming error in our own code we abort, but if the user made one we try to go on.

Reply:
That makes sense. These are all "programmer error" asserts. I've replaced it with plain assert().